storage/innobase/log/log0log.cc

```diff
@@ -516,17 +516,14 @@ log_t::resize_start_status log_t::resize_start(os_offset_t size) noexcept
         resize_target= size;
         resize_buf= static_cast<byte*>(ptr);
         resize_flush_buf= static_cast<byte*>(ptr2);
+        start_lsn= get_lsn();
+
         if (is_pmem())
-        {
           resize_log.close();
-          start_lsn= get_lsn();
-        }
         else
-        {
-          memcpy_aligned<16>(resize_buf, buf, (buf_free + 15) & ~15);
           start_lsn= first_lsn +
-            (~lsn_t{write_size - 1} & (write_lsn - first_lsn));
-        }
+            (~lsn_t{write_size - 1} &
+             (lsn_t{write_size - 1} + start_lsn - first_lsn));
```

Comment:
Here we are changing the start_lsn from write_lsn (aligned down) to log_sys.lsn (aligned up). If I am right, this change is independent to the other issue/bug that is being resolved by this patch. 

1. I could not understand what issue is being resolved by this part. Is there any issue if the current write_lsn is used ? Is it solving anything or some attempted improvement ?
2.  This is setting the start lsn after resize to a possible future LSN. Is this causing the resize to wait in the test ?

> This may block until some InnoDB persistent data is changed.

AFAICS, this change would cause behaviour change in a negative way.

Reply:
Before I came up with this simple solution, I had added a data member to `log_sys` that would be a copy of the current local variable `write_buf` in `log_t::write_buf()`. It is this buffer that should form the ‘seed’ of `resize_buf`. However, that would not be enough; the test `innodb.log_file_size_online` would still fail, rather rarely. The problem is that  the old `ib_logfile0` may have wrapped around an odd number of times, and therefore the sequence bit at the end of a mini-transaction would be 0 and not the initial value 1.

In the non-memory-mapped I/O, there are some constraints around `log_sys.buf_free`. To keep things simple, we align the `start_lsn` so that we can write the same blocks of data (except for the sequence bit) to both `log_sys.log` and `log_sys.resize_log` . The formula here used to be able to choose a `start_lsn` that is smaller than the current `log_sys.lsn`. If the log resizing completes with such a LSN, then recovery could start parsing the resized log from such an LSN. That would be all fine and good if the ‘seed’ buffer contained sequence bits that are 1. However, if the sequence bits are 0, recovery would determine that it reached the end of the circular log, and stop right at the checkpoint. This was what motivated me to file #3476; the failure was not at all obvious.

If we wanted to adjust the sequence bits so that recovery can work, we’d have to add even more bookkeeping, just to satisfy a hard-to-test condition of a rarely used feature. We’d basically have to remember each "log block" or mini-transaction boundary, because the log can only be scanned forward, not backwards. The logic regarding adjusting the sequence bit (the assignments `resize_buf[s]= 1` in `log_t::resize_write()`) would only be enabled after this code has been executed. I would not want to execute that code unconditionally, just in case someone is going to initiate a log file size change on an idle or read-only server.

Any added bookkeeping in this area would not come for free. In bf0b82d24ba5cfa6d3cf42fba501ba4a57f266ff I spent a lot of effort to optimize the cache line usage, so that two 64-byte x86-64 cache lines pointing to `log_sys` are being used in a sensible way during `mtr_t::commit()`. If we added more data fields to a busy code path, a serious performance regression could occur.

Reply:
By the way, I had tried to justify this change in the commit message  e63dae8a15ad52f127818bc0c3a4e4b5f6a51358. I will try to make it even clearer.

Reply:
Yes, in [MDEV-14425](https://jira.mariadb.org/browse/MDEV-14425) (685d958e38b825ad9829be311f26729cccf37c46) the 31-bit `LOG_BLOCK_HDR_NO` was shrunk to 1 bit. That bit will flip if and only if the the sequence byte (right before the CRC-32C of the mini-transaction bytes) is at a lower file offset (near `log_sys.START_OFFSET`) than the start of the mini-transaction. That does make the value of the bit dependent on the `innodb_log_file_size`.

We can (and actually during `SET GLOBAL innodb_log_file_size` as well as in `mariadb-backup --backup` already do) replicate log records into an *append-only* fashion, always setting the sequence bit as 1. I specifically chose the initial value as 1 so that we can more conveniently detect the end of the log when the `ib_logfile0` has initially been filled with NUL bytes.

I would be very reluctant to change a file format.

To better understand the problem scenario, you can revert this code change and run
```bash
while ./mtr --rr --parallel=auto innodb.log_file_size_online{,,,,,,,,,,,,,,}; do :; done
```
and finally check the failure with `rr replay`. That is how I came up with this solution.

Reply:
@dr-m explained to me that the problem is smaller in magnitude than I perceived. That is we already generate a different buffers at mtr commit for resize log and the sequence BIT is set there correctly. This double copy may not look very optimal but should be fine only during resize time.

So, the only issue left was the log that were already written by mtrs but not yet flushed, i.e. write_lsn -> log_sys.lsn. The current patch is trying to get rid of this problem by trying to set the start LSN for resized log to log_sys.lsn instead. However, due to alignment requirement we need to go either ahead or behind log_sys LSN. Going behind would need addressing the incorrect sequence BIT and going ahead (the current patch) could lead to indefinite wait when server is idle.

So, it seems ...

Either 

- we accept the apparent hang of the SET GLOBAL in idle server and document it. User needs to have some concurrent operation to let the SET GLOBAL complete.

Or
- make more code changes to address it.

Possible solutions seems to be to parse the redo logs from write_lsn -> log_sys LSN and populate the resize_buf with correct sequence BIT or start from log_sys.lsn rounded up and device dummy mtrs to progress the LSN.

Reply:
To avoid a hang of `SET GLOBAL innodb_log_file_size`, `innodb_log_file_size_update()` can write enough dummy log records so that the LSN will advance enough. There is an `#if 0` section code `log_pad()` for that purpose. Recovery can deal with such dummy records.

Reply:
984606d74705e1a4c67945531edeff81692f403d addresses this by making `innodb_log_file_size_update()` write redundant `FILE_CHECKPOINT` records if needed. The changes to the test `innodb.log_file_size_online` are smaller, so that it is likely that the first `SET GLOBAL innodb_log_file_size` would hang if this tweak were not present. The tweak (as well as the hang) was validated by running 3×2×10 rounds of that test. At least 1 of the 6 concurrent test runs would typically hang when `mysql-test/var` is linked to `/run/user/$UID/var`.